add tests with gh actions

djay opened the pull request.

djay committed:
add base actions setup

djay committed:
try with simpler buildout

djay committed:
install dependencies

djay committed:
better dep

djay committed:
more version pins

djay committed:
fix unittest

davisagli commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,79 @@
+          - plone-version: 5.1
+            python-version: 3.7
+          - plone-version: 5.1
+            python-version: 3.8
```
I think you could add the 5.2 + python 3.8 combo using `include` instead of excluding the others: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-new-combinations

thyarles reviewed (commented):
The changes looks good for me. I'll try to run it later.

djay committed:
fix unittest2

djay committed:
right versions to build

djay committed:
flake8 fixes

djay committed:
start of tests

djay committed:
fix versions

djay committed:
run right tests

djay committed:
version pins for 5.0

djay committed:
fix initial test setup

djay commented:
@davisagli @thyarles got any idea how to setup 2 plone sites in a zope test layer? Or some other bright idea to simulate saml2 auth exchange?

djay committed:
try to setup 2 plone sites to test login

djay committed:
try to improve the readme instructions

djay committed:
try setting up a 2nd site

djay committed:
confirm using plan zope is not enough. needs two plone sites

djay committed:
fix up docs to show that urls for entity names will work

wesleybl commented:
@djay I don't see GitHub Actions in this PR. Can you enable it in the repository settings?

djay committed:
start to switch to using mock-idp for testing

wesleybl committed:
Run Python 2 Jobs on the python:2.7.18-buster image

wesleybl committed:
Don't use container in Python 3

wesleybl committed:
Not use setup-python in Python 2

wesleybl committed:
Don't use container in Python 3

wesleybl committed:
Don't use sudo in Python 2

sudo command does not exist in python:2.7.18-buster

wesleybl committed:
Determines whether commands will be executed with sudo based on the
Python version

sudo command does not exist in python:2.7.18-buster

wesleybl committed:
Fixes logic that defines whether sudo will be used

It is in Python 3 that we must use sudo

wesleybl committed:
Fix version in test-5.2.x.cfg

wesleybl committed:
Move repeated configurations in test-x.x.x.cfg to buildout.cfg

wesleybl committed:
Move parts in buildout.cfg to base.cfg

wesleybl committed:
Does not install python-dev package

Avoid the error:

E: Package 'python-dev' has no installation candidate

wesleybl committed:
Does not install python-saml package

Avoid the error in Python 3:

SyntaxError: Missing parentheses in call to 'print'. Did you mean print("Warning : cflag %s skipped" % flag)?

wesleybl committed:
Add requirements.txt

wesleybl committed:
Upgrade pip

djay commented:
@wesleybl so basically my plan plan originally was to use two plone sites and test against each other. One as IdP and one as a SP. but getting that to work with plone.app.test worked out too hard.
So instead I figured just test it as a SP as thats the most common case and test it against mock-idp which seems pretty simple. One thing to overcome is that dm.saml2 requires the publishing of a metadata file and mock-idp doesn't do this so a metafile will need to be generated manually and published somehow (maybe even a file link would work. if not a simple http server). After that I think it should not so hard.
No need to keep tests in doctest format. only reason I did that is it makes it a bit easier to test testbrowser output.
Anyway. thanks for giving it a go and if it proves too hard I'm sure I will find time eventually.

wesleybl committed:
pep517 = 0.12.0 in Python 2

pep517 > 0.12.0 is not compatible with Python 2.

wesleybl committed:
Remove pinns

wesleybl committed:
Remove pinns

wesleybl commented:
> So instead I figured just test it as a SP as thats the most common case and test it against mock-idp which seems pretty simple.

@djay `mock-idp` doesn't work with Python 2:

https://github.com/collective/collective.saml2/actions/runs/9509927169/job/26213629926#step:9:776

Actually, for now I'm more interested in Python 2

wesleybl committed:
Change index in Plone 5.0

wesleybl committed:
Upgrade setuptools and zc.buildout in Plone 5.0

wesleybl committed:
Fix pinn

wesleybl committed:
Remove pinns

wesleybl committed:
zope.app.component = 3.9.3 in Plone 5.1

Avoid:

Version and requirements information containing zope.app.container:
  [versions] constraint on zope.app.container: 3.9.2
  Requirement of zope.app.component: zope.app.container>=4.0
While:
  Installing instance.
Error: The requirement ('zope.app.container>=4.0') is not allowed by
your [versions] constraint (3.9.2)

wesleybl committed:
five.formlib = 1.0.4 in Plone 5.0

Avoid:

Version and requirements information containing zope.formlib:
  [versions] constraint on zope.formlib: 4.0.6
  Requirement of zope.app.component: zope.formlib
  Requirement of zope.formlib: zope.datetime
  Requirement of zope.formlib: zope.traversing
  Requirement of zope.formlib: zope.security
  Requirement of zope.formlib: zope.schema>=3.5.1
  Requirement of zope.formlib: zope.publisher
  Requirement of zope.formlib: zope.lifecycleevent
  Requirement of zope.formlib: zope.interface
  Requirement of zope.formlib: zope.i18nmessageid
  Requirement of zope.formlib: zope.i18n
  Requirement of zope.formlib: zope.event
  Requirement of zope.formlib: zope.component
  Requirement of zope.formlib: zope.browserpage>=3.11.0
  Requirement of zope.formlib: zope.browser>=1.1
  Requirement of zope.formlib: pytz
  Requirement of zope.formlib: setuptools
  Requirement of five.formlib: zope.formlib>=4.4
  Requirement of plone.app.collection: zope.formlib
While:
  Installing instance.
Error: The requirement ('zope.formlib>=4.4') is not allowed by your
[versions] constraint (4.0.6)

wesleybl committed:
signxml = 2.10.1 in Plone 5.2

Avoid:

Version and requirements information containing cryptography:
  [versions] constraint on cryptography: 3.3.2
  Requirement of signxml: cryptography>=3.4.8
  Requirement of pyOpenSSL>=19.0.0: cryptography<43,>=41.0.5
While:
  Installing test.
Error: The requirement ('cryptography>=3.4.8') is not allowed by your
[versions] constraint (3.3.2)

djay commented:
@wesleybl good point.  But it's not an easy thing to test. Maybe there is another IDP code that is easy to set up that does support python 2. 
Or just run the test only in python 3 and hope it still works in 2. Support will have to be dropped for 2 at some point anyway.

wesleybl committed:
Don't pinn pep517

The pinn was made in buildout.plonetest

wesleybl committed:
Don't use mock-idp for now

wesleybl committed:
readme-renderer = 28.0 in Plone 5.1

Newer versions are not compatible with Python 2

wesleybl committed:
lxml = 4.0.0 in Plone 5.0

Avoid:

ImportError: lxml.etree does not export expected C function adoptExternalDocument

wesleybl committed:
packaging = 20.9 in Plone 5.1

Newer versions are not compatible with Python 2.

wesleybl committed:
Fix bindings import in Python 3

wesleybl committed:
Omelette

wesleybl committed:
Fix flake8

wesleybl committed:
Test with Plone 4.3

wesleybl committed:
Move doctest to tests folder

wesleybl committed:
Add install test

wesleybl commented:
@djay I configured Plone 6.0 with Python 3.8 to authenticate with

https://mocksaml.com/

but an error occurred. The server went down, without a traceback. I debugged it and saw that the error is on line 308 of the file `dm/saml2/signature.py` .The line in question is:

```python
addIDs(doc_tree.getroot(), ['ID'])
```

So it looks like it doesn't work entirely well with Python 3

wesleybl committed:
five.formlib = 2.2 in Plone 5.2 whit Plone 2

wesleybl commented:
In Plone 5.2 with Python2, and configured with

https://mocksaml.com

I get the error:

```bash
AttributeError: 'NoneType' object has no attribute 'startswith'
2024-06-14 16:31:04,856 ERROR   [Zope.SiteErrorLog:252][waitress-1] 1718393464.860.0667113140391 http://localhost:8080/Plone/acl_users/saml2sp/authenticate
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module dm.zope.saml2.spsso.spsso, line 102, in authenticate
  Module dm.zope.saml2.role, line 148, in deliver
  Module pyxb.binding.basis, line 560, in toxml
  Module pyxb.binding.basis, line 533, in toDOM
  Module dm.saml2.signature, line 311, in finalize
  Module dm.saml2.signature, line 154, in sign
  Module dm.zope.saml2.authority, line 423, in sign
  Module dm.zope.saml2.authority, line 317, in _get_signature_context
  Module dm.zope.saml2.authority, line 323, in _add_sign_keys
  Module dm.zope.saml2.authority, line 440, in _make_absolute
  Module posixpath, line 54, in isabs
AttributeError: 'NoneType' object has no attribute 'startswith'
```

djay commented:
@wesleybl nice idea using that service. Can you raise a seperate bug report?

djay commented:
@wesleybl looks like it might be related to creating of a local certificate. This might be missing from the test setup

wesleybl commented:
> @wesleybl looks like it might be related to creating of a local certificate. This might be missing from the test setup

@djay It appears that https://mocksaml.com/ needs the authentication request to be signed. But I don't know if `dm.zope.saml2` supports this.

I haven't tried the test yet. I was trying on the local website.
